Change the splitting method used to extract option values

gaudryc opened the pull request:
Change the splitting method used to extract option values to prevent the problem causing a shift of value (Helper::SplitString compresses empty strings).

gaudryc committed:
Change the splitting method used to extract option values to prevent the
problem causing a shift of value (Helper::SplitString compresses empty
strings).

gizmocuz commented:
Please fix the original function

gaudryc commented:
Do you want me to change the SplitString function ?

gizmocuz commented:
Yes please, it is used in a lot of places, so if it is wrong, please correct it

gaudryc commented:
Do you want the SplitString method to behave like boost::split(...,..., compress off) ?

gaudryc commented:
Because the original SplitString ignores empty strings like boost::split with compress on.

So if we change SplitString behaviour we have to be sure that all 145 callers of this method want to have empty strings also.

gizmocuz commented:
I dont know what is wrong with the original function, but it seems you had a problem with it
It splits a string... if it needs to do more, maybe keep the boost change ;)

gaudryc commented:
I'll make you a draw in 1 hour ;-)

gizmocuz commented:
Please take your time.... ;)

gaudryc commented:
So...

What StringSplit function does using pipe delimiter ?
1. StringSplit("", "|", strarray) => []
2. StringSplit("|", "|", strarray) => []
3. StringSplit("value0", "|", strarray) => [value0]
4. StringSplit("value0|value1", "|", strarray) => [value0,value1]
5. StringSplit("|value1|value2", "|", strarray) => [value1,value2]
6. StringSplit("|value1|", "|", strarray) => [value1]
7. StringSplit("|value1|||value4", "|", strarray) => ["value1","value4"]
In examples 1, 2, 5, 6 and 7 you can not know that the string to split contains "empty" values. You can only have not empty values.

What boost::split function does using pipe delimiter and boost::token_compress_on ?
1. split(strarray, "", is_any_of("|"), token_compress_on) => [""]
2. split(strarray, "|", is_any_of("|"), token_compress_on) => ["",""]
3. split(strarray, "value0", is_any_of("|"), token_compress_on) => ["value0"]
4. split(strarray, "value0|value1", is_any_of("|"), token_compress_on) => ["value0","value1"]
5. split(strarray, "|value1|value2", is_any_of("|"), token_compress_on) => ["","value1","value2"]
6. split(strarray, "|value1|", is_any_of("|"), token_compress_on) => ["","value1",""]
7. split(strarray, "|value1|||value4", is_any_of("|"), token_compress_on) => ["","value1","value4"]
Now, in examples 1, 2, 5, 6 and 7 you know that the string to split contains "empty" values.
But in example 7 you can not know that the string to split does not contain value2 and value3.

What boost::split function does using pipe delimiter and boost::token_compress_off ?
1. split(strarray, "", is_any_of("|"), token_compress_off) => [""]
2. split(strarray, "|", is_any_of("|"), token_compress_off) => ["",""]
3. split(strarray, "value0", is_any_of("|"), token_compress_off) => ["value0"]
4. split(strarray, "value0|value1", is_any_of("|"), token_compress_off) => ["value0","value1"]
5. split(strarray, "|value1|value2", is_any_of("|"), token_compress_off) => ["","value1","value2"]
6. split(strarray, "|value1|", is_any_of("|"), token_compress_off) => ["","value1",""]
7. split(strarray, "|value1|||value4", is_any_of("|"), token_compress_off) => ["","value1","","","value4"]
Now, in all the examples, you know what values are empty.

So I think that changing the function StringSplit needs to be done very carefully because its behaviour is really different. It is not clean but it seems that it works for now.

To change the function StringSplit we should proceed step by step (each time it is called) to add some defensive code to handle empty values and protect the code against segmentation faults.

gizmocuz commented:
Okey, i merged the pull request ;) thanks for looking into this

gaudryc commented:
Are you waiting for me to do something about the StringSplit function ?
I can make a review to replace it with boost::split where it can be done safely.

gizmocuz commented:
No, i merged your changes from this pull request...

gaudryc commented:
OK so I let it like this for now.

gizmocuz commented:
Perfect, thanks!
